scripts/packaging/createupmpackages.ps1

```diff
@@ -0,0 +1,192 @@
+.PARAMETER PackageVersion
+    What version of the artifacts should we build? If unspecified, the highest
+    git tag pointing to HEAD is searched. If none is found, an error is reported.
+.PARAMETER OutputTarget
+    What is the target for the artifacts? To the Official server ("official"), test server ("test") or local folder ("local")?
+#>
+param(
+    [ValidatePattern("^\d+\.\d+\.\d+")]
+    [string]$NodejsVersion = "12.18.0",
+    [string]$OutputDirectory = ".\artifacts\upm",
+    [ValidatePattern("^\d+\.\d+\.\d+-?[a-zA-Z0-9\.]*$")] # todo - format of d.d.d[-preview.0-9.0-9]
+    [string]$PackageVersion,
+    [string]$OutputTarget = "local"
+)
+
+if (-not $PackageVersion) {
+        throw "Unknown package version. Please specify -PackageVersion when building."
+}
+
+if ($OutputTarget -eq "local") {
+    if (-not (Test-Path $OutputDirectory -PathType Container)) {
+        New-Item $OutputDirectory -ItemType Directory | Out-Null
+    }
+    $OutputDirectory = Resolve-Path "$(Get-Location)\$OutputDirectory"
+}
+
+$projectRoot = Resolve-Path "$(Get-Location)\..\.." 
+$scriptPath = "$projectRoot\scripts\packaging"
+
+$scope = "com.microsoft.mixedreality"
+$product = "toolkit"
+
+# This hashtable contains mappings of the packages (by name) to the folder which contains
+# the package contents.
+#
+# The keys of this hashtable will be combined with the scope and product to create the 
+# final name of the package (for example, com.microsoft.mixedreality.toolkit.foundation 
+# will use the packaging file contained in the folder associated with the foundation key).
+#
+# Note that capitalization below in the key itself is significant. Capitalization
+# in the values is not significant.
+#
+# These paths are projectRoot relative.
+$packages = [ordered]@{
+    "foundation" = "Assets\MRTK";
+    # providers
+    "leapmotion" = "Assets\MRTK\Providers\LeapMotion";
+    "oculus.xrplugin" = "Assets\MRTK\Providers\Oculus\XRSDK"
+    "openvr" = "Assets\MRTK\Providers\OpenVR";
+    "unityar" = "Assets\MRTK\Providers\UnityAR";
+    "windows" = "Assets\MRTK\Providers\Windows";
+    "wmr" = "Assets\MRTK\Providers\WindowsMixedReality\XR2018";
+    "wmr.xrplugin" = "Assets\MRTK\Providers\WindowsMixedReality\XRSDK";
+    "wmr.shared" = "Assets\MRTK\Providers\WindowsMixedReality\Shared";
+    "xrplugin" = "Assets\MRTK\Providers\XRSDK";
+    # extensions
+    "handphysicsservice" = "Assets\MRTK\Extensions\HandPhysicsService";
+    "losttrackingservice" = "Assets\MRTK\Extensions\LostTrackingService";
+    "scenetransitionservice" = "Assets\MRTK\Extensions\SceneTransitionService";
+    # other packages
+    "tools" = "Assets\MRTK\Tools";
+    "testutilties" = "Assets\MRTK\Tests\TestUtilities";
+    "examples" = "Assets\MRTK\Examples";
+}
+
+# Acquire and unpack node.js
+$archiveName = "node-v$NodejsVersion-win-x64"
+$archiveFile = ".\$archiveName.zip"
+$downloadUri = "https://nodejs.org/dist/v$NodejsVersion/$archiveFile"

```

Comment:
Are we able to use the Azure Pipelines npm task instead of needing downloading something from a URL in powershell?

https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/package/npm

We might even be able to use task-encoded registry info instead of the .npmrc files, which could help simplify the packaging step:

>You can either commit a .npmrc file to your source code repository and set its path or select a registry from Azure Artifacts.

Reply:
I'll look into that

Reply:
For consistency with the nuget packaging script, i would like this one to behave similarly. i will check for a pre-installed node.js and only fetch the package if not found,

Reply:
added a check for a previously installed version. only download and unpack if not on the path

Reply:
Still feel slightly odd about not using the built-in task, and I'm not sure I care much about having consistency with a script that we're phasing out. Not blocking on this, but wanted to comment again!